datalad/interface/rerun.py

```diff
@@ -0,0 +1,300 @@
+            constraints=EnsureStr() | EnsureNone()),
+        onto=Parameter(
+            metavar="base",
+            args=("--onto",),
+            doc="""start point for rerunning the commands. If not
+            specified, commands are executed at HEAD. This option can
+            be used to specify an alternative start point, which will
+            be checked out with the branch name specified by --branch
+            or in a detached state otherwise. As a special case, an
+            empty value for this option means to use the commit
+            specified by --since.""",
+            constraints=EnsureStr() | EnsureNone()),
+        # TODO
+        # --list-commands
+        #   go through the history and report any recorded command. this info
+        #   could be used to unlock the associated output files for a rerun
+    )
+
+    @staticmethod
+    @datasetmethod(name='rerun')
+    @eval_results
+    def __call__(
+            revision="HEAD",
+            since=None,
+            dataset=None,
+            branch=None,
+            message=None,
+            onto=None):
+
+        ds = require_dataset(
+            dataset, check_installed=True,
+            purpose='rerunning a command')
+
+        lgr.debug('rerunning command output underneath %s', ds)
+
+        from datalad.tests.utils import ok_clean_git
+        try:
+            ok_clean_git(ds.path)
+        except AssertionError:
+            yield get_status_dict(
+                'run',
+                ds=ds,
+                status='impossible',
+                message=('unsaved modifications present, '
+                         'cannot detect changes by command'))
+            return
+
+        err_info = get_status_dict('run', ds=ds)
+        if not ds.repo.get_hexsha():
+            yield dict(
+                err_info, status='impossible',
+                message='cannot rerun command, nothing recorded')
+            return
+
+        if branch and branch in ds.repo.get_branches():
+            yield get_status_dict(
+                "run", ds=ds, status="error",
+                message="branch '{}' already exists".format(branch))
+            return
+
+        root = False
+        if since is None:
+            revrange = "{}^..{}".format(revision, revision)
+        elif since.strip() == "":
+            revrange = revision
+            root = True
+        else:
+            revrange = "{}..{}".format(since, revision)
+
+        if ds.repo.repo.git.rev_list("--merges", revrange, "--"):
+            yield get_status_dict(
+                "run", ds=ds, status="error",
+                message="cannot rerun history with merge commits")
+            return
+        revs = ds.repo.repo.git.rev_list("--reverse", revrange, "--").split()
+
+        do_checkout = branch
+        if onto is not None:
+            if onto.strip() == "":
+                ## An empty argument means go to the parent of the
```

Comment:
what is your convention for having `##` comment prefix?  pycharm is screaming at me for being non PEP8 compliant here.

Reply:
sorry, elisp comment conventions are leaking into my python.

I've been catching a good number pre-commit (there were many more :).
Will fix.